### PR TITLE
WT-11845 Fix transaction visibility issue with truncate

### PR DESCRIPTION
**Pull request by jiechenbo**
This ticket fixes a transaction visibility issue when performing fast truncate. The truncate does not account for the transaction ids that occur before the truncate transaction which can lead to records being removed unnoticed. To account for this, we use the snap_min as an indicator if the truncate transaction sees any other possible concurrent transactions that may happen.

### CODE REVIEW

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Review by agorrod — CHANGES_REQUESTED** [review]
Since we have understood the root cause of the failure, is it possible to create a more predictable reproducer?
I don't know if it is predictable enough that a Python reproducer can be constructed, if not the C++ test suite seems like a good candidate for a focused reproducer.

**Comment by jeremythorp** [comment]
> Since we have understood the root cause of the failure, is it possible to create a more predictable reproducer? I don't know if it is predictable enough that a Python reproducer can be constructed, if not the C++ test suite seems like a good candidate for a focused reproducer.

I think we'll be able to create a simple, predictable reproducer, and @clarissecheah + @ajmorton are on the case!

**Comment by jeremythorp** [comment]
> > Since we have understood the root cause of the failure, is it possible to create a more predictable reproducer? I don't know if it is predictable enough that a Python reproducer can be constructed, if not the C++ test suite seems like a good candidate for a focused reproducer.
> 
> I think we'll be able to create a simple, predictable reproducer, and @clarissecheah + @ajmorton are on the case!

@clarissecheah & @ajmorton have created a simple Python reproducer, which is awesome, and they have added it into this PR.

**Review by agorrod — COMMENTED** [review]
The fix here looks good to me, could you add some protection against re-introducing it to a dist script, something like would satisfy me even though I know it isn't very complete:

```
$ cat dist/s_visibility_checks 
#!/bin/bash

t=__wt.$$
trap 'rm -f $t' 0 1 2 3 13 15

cd ..

# Check to see if a time aggregate is used in a straight visibility check.
# It should never be - the "snap_min" visibility check should be used instead.
found=`grep -rnI "txn_visible.*addr.ta" src/`
if [ ! -z "$found" ]; then
	echo "Found invalid usage(s) of visibility check and a time aggregate window"
	echo "$found"
	exit 1
fi

exit 0
```

**Comment by jiechenbo** [comment]
Great idea @agorrod , I added the script to the PR. I was thinking of improving the script but realised it is already quite simple and powerful, I can't find anything to improve, let me know what other great additions you were thinking of adding to the script.

**Comment by agorrod** [comment]
> Great idea @agorrod , I added the script to the PR. I was thinking of improving the script but realised it is already quite simple and powerful, I can't find anything to improve, let me know what other great additions you were thinking of adding to the script.

I think making the script better involves making it significantly more sophisticated - parsing out all calls to the visibility functions and their parameters and then check to see whether the parameter is of the right type.
I tried extending the script to pull out the visibility function and a subsequent line (so it could catch split parameters most of the time), but even that got more complex and didn't yield immediate benefit.

I'd be happy for those extensions to be future work on the backlog (or not at all until they are needed).

**Review by agorrod — APPROVED** [review]
LGTM

File: src/btree/bt_delete.c [hunk 1]
```diff
@@ -32,15 +33,15 @@
...
- * marked deleted, a pointer to the WT_REF was saved in the deleting session's transaction list and
- * the delete is unrolled by resetting the WT_REF_DELETED state back to WT_REF_DISK. However, if the
- * page has been instantiated by some reading thread, that's not enough; each individual row on the
- * page must have the delete operation reset. If the page split, the WT_UPDATE lists might have been
- * saved/restored during reconciliation and appear on multiple pages, and the WT_REF stored in the
- * deleting session's transaction list is no longer useful. For this reason, when the page is
- * instantiated by a read, a list of the WT_UPDATE structures on the page is stored in the
- * WT_PAGE_MODIFY.inst_updates field. That way the session resolving the delete can find all
+ * There's an additional complication to support transaction rollback of the page delete. When the
+ * age was marked deleted, a pointer to the WT_REF was saved in the deleting session's transaction
```
Comment (jeremythorp) [inline src/btree/bt_delete.c:37]: There's a small typo here: 'age' should be 'page'.

File: src/btree/bt_delete.c [hunk 2]
```diff
@@ -14,14 +14,15 @@
...
+ * The way session truncate works is it explicitly reads the first and last pages of the truncate
  * range, then walks the tree with a flag so the tree walk code skips reading eligible pages within
  * the range and instead just marks them as deleted, by changing their WT_REF state to
  * WT_REF_DELETED. Pages ineligible for this fast path ("fast-truncate" or "fast-delete") include
- * pages already in the cache, having overflow items, containing prepared values, or belonging to
- * FLCS trees. Ineligible pages are read and have their rows updated/deleted individually
- * ("slow-truncate"). The transaction for the delete operation is stored in memory referenced by the
- * WT_REF.page_del field.
+ * pages that are already in the cache and can not be evicted, records in the pages that are
+ * not visible to the transaction, page containing overflow items, pages containing prepared values,
```
Comment (jeremythorp) [inline src/btree/bt_delete.c:22]: 'page' here probably should be 'pages'

**Review by jeremythorp — APPROVED** [review]
LGTM!

**Comment by jiechenbo** [comment]
evergreen merge

**Pull request merged**